ci/remote-watch.py

```diff
@@ -0,0 +1,201 @@
+            os.kill(os.getppid(), signal.SIGTERM)
+        else:
+            os.kill(os.getppid(), signal.SIGTERM)  # apparently this is needed
+            time.sleep(timeout)
+            os.kill(0, signal.SIGKILL)
+    except OSError as ex:
+        if ex.errno not in (errno.EBADF, errno.ESRCH):
+            raise
+        logger.error("Kill error %s: %s", ex.errno, ex.strerror)
+        result = ex.errno
+    return result
+
+
+def yield_poll_schedule():
+    schedule = [0, 5, 5, 10, 20, 40, 40] + [60] * 5 + [120] * 10 + [300, 600]
+    for item in schedule:
+        yield item
+    while True:
+        yield schedule[-1]
+
+
+def monitor():
+    ci = get_current_ci() or ""
+    (ref, remote, expected_sha) = git_branch_info()
+    expected_line = "{}\t{}".format(expected_sha, ref)
+    commit_msg = git("show", "-s", "--format=%B", "HEAD^-")
+    keep_alive = False
+    for match in re.findall("(?m)^([# ]*CI_KEEP_ALIVE(:(.*))?)$", commit_msg):
+        if not match[2].strip():
+            keep_alive = True
+        else:
+            for ci_name in match[2].split(","):
+                if ci_name.strip().lower() == ci.lower():
+                    keep_alive = True
+    logger.info("Monitoring %s (%s) for changes in %s: %s",
+                remote, git("ls-remote", "--get-url", remote),
+                ref, expected_line)
+    prev_ignored_line = None
+    terminate = False
+    for to_wait in yield_poll_schedule():
+        time.sleep(to_wait)
+        status = 0
+        line = None
+        try:
+            line = git("ls-remote", "--exit-code", remote, ref)
+        except subprocess.CalledProcessError as ex:
+            status = ex.returncode
+        if status == 2:
+            terminate = True
+            logger.info("Terminating job as %s has been deleted on %s: %s",
+                        ref, remote, expected_line)
+            break
+        elif status != 0:
+            logger.error("Error %d: unable to check %s on %s: %s",
+                         status, ref, remote, expected_line)
+        elif line == expected_line:
+            pass  # everything good
+        elif keep_alive:
+            if prev_ignored_line != line:
+                logger.info("Not terminating job even though %s has changed "
+                            "on %s due to keep-alive on: %s",
+                            ref, remote, expected_line)
+                prev_ignored_line = line
+        else:
+            terminate = True
+            logger.info("\n".join(
+                [
+                    "Terminating job as %s has changed on %s",
+                    "    from:\t%s",
+                    "    to:  \t%s",
+                ]), ref, remote, expected_line, line)
+            break
+    if terminate:
+        result = terminate_my_process_group()
+    return result
+
+
+def main(program, *args):
+    p = argparse.ArgumentParser()
+    p.add_argument("--skip_repo", action="append", help="Repo to exclude.")
```

Comment:
Can we hard code this as a constant in this file?

Reply:
I was trying not to make the script Ray-specific... it might be useful elsewhere?